dcm4che-imageio/src/main/java/org/dcm4che3/imageio/plugins/dcm/DicomImageReader.java

```diff
@@ -587,22 +618,83 @@ private void readMetadata() throws IOException {
                 pixelDataLength = dis.length();
                 if (pixelDataLength == -1)
                     epdiis = new EncapsulatedPixelDataImageInputStream(dis, ds.getInt(Tag.NumberOfFrames, 1));
+            } else {
+                Attributes postPixelData = dis.readDataset(-1, -1);
+                ds.addAll(postPixelData);
             }
             setMetadata(new DicomMetaData(fmi, ds));
             return;
         }
         if (iis == null)
             throw new IllegalStateException("Input not set");
 
-        @SuppressWarnings("resource")
         DicomInputStream dis = new DicomInputStream(new ImageInputStreamAdapter(iis));
         dis.setIncludeBulkData(IncludeBulkData.URI);
         dis.setBulkDataDescriptor(BulkDataDescriptor.PIXELDATA);
         dis.setURI("java:iis"); // avoid copy of pixeldata to temporary file
         Attributes fmi = dis.readFileMetaInformation();
-        Attributes ds = dis.readDataset(-1, -1);
-        initPixelData(ds);
+        Attributes ds = dis.readDataset(-1, Tag.PixelData);
+        if( dis.tag() == Tag.PixelData ) {
+            pixelDataVR = dis.vr();
+            pixelDataLength = dis.length();            
+        } else {
+            Attributes postPixelData = dis.readDataset(-1, -1);
+            ds.addAll(postPixelData);
+        }
         setMetadata(new DicomMetaData(fmi, ds));
+        initPixelDataIIS(dis);
+    }
+
+    /** Initializes the pixel data reading from an image input stream */
+    private void initPixelDataIIS(DicomInputStream dis) throws IOException {
+        if( pixelDataLength==0 ) return;
+        if( pixelDataLength>0 ) {
+            pixelData = new BulkData("pixeldata://", dis.getPosition(), dis.length(),dis.bigEndian());
+            metadata.getAttributes().setValue(Tag.PixelData, pixelDataVR, pixelData);
+            return;
+        }
+        dis.readItemHeader();
+        byte[] b = new byte[dis.length()];
+        dis.readFully(b);        
+
+        long start = dis.getPosition();
+        pixelDataFragments = new Fragments(pixelDataVR, dis.bigEndian(), frames);
+        pixelDataFragments.add(b);
+        
+        generateOffsetLengths(pixelDataFragments, frames,b, start);
+    }
+
+    /** Creates an offset/length table based on the frame positions */
+    public static void generateOffsetLengths(Fragments pixelData, int frames, byte[] basicOffsetTable, long start) {
+        long lastOffset = 0;
+        BulkData lastFrag = null;
+        for(int frame=0; frame<frames; frame++) {
+            long offset = frame>0 ? 1 : 0;
+            int offsetStart = frame*4;
+            if( basicOffsetTable.length>=offsetStart+4 ) {
+                offset = ByteUtils.bytesToIntLE(basicOffsetTable, offsetStart);
+                if( offset!=1 ) {
```

Comment:
Allow 1 values to indicate unknown yet basic offset table values - that allows writing an image, then going back and fixing up the basic offset table, will be used in a later code review.